liblocation/src/main/java/com/mapbox/android/core/location/MapboxFusedLocationEngineImpl.java

```diff
@@ -0,0 +1,160 @@
+import static com.mapbox.android.core.location.Utils.isBetterLocation;
+
+/**
+ * Mapbox replacement for Google Play Services Fused Location Client
+ * <p>
+ * Note: optimize engine logic for availability
+ */
+class MapboxFusedLocationEngineImpl extends AndroidLocationEngineImpl implements LocationListener {
+  private static final String TAG = "MapboxLocationEngine";
+
+  private LocationEngineCallback<LocationEngineResult> callback;
+  private Location currentBestLocation;
+
+  MapboxFusedLocationEngineImpl(@NonNull Context context) {
+    super(context);
+    this.callback = null;
+  }
+
+  @NonNull
+  @Override
+  LocationListener createListener(final LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = callback;
+    return this;
+  }
+
+  @Nullable
+  @Override
+  public LocationListener removeLocationListener(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = null;
+    return super.removeLocationListener(callback);
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    Location bestLastLocation = getBestLastLocation();
+    if (bestLastLocation != null) {
+      callback.onSuccess(LocationEngineResult.create(bestLastLocation));
+    } else {
+      callback.onFailure(new Exception("Last location unavailable"));
+    }
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull LocationListener listener,
+                                     @Nullable Looper looper) throws SecurityException {
+    super.requestLocationUpdates(request, listener, looper);
+
+    // Start network provider along with gps
+    if (shouldStartNetworkProvider(request.getPriority())) {
+      try {
+        locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
+                request.getInterval(), request.getDisplacemnt(),
+                listener, looper);
+      } catch (IllegalArgumentException iae) {
+        iae.printStackTrace();
+      }
+    }
+
+    // Fetch best last location
+    Location bestLastLocation = getBestLastLocation();
+    if (bestLastLocation != null) {
+      onLocationChanged(bestLastLocation);
+    }
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull PendingIntent pendingIntent) throws SecurityException {
+    super.requestLocationUpdates(request, pendingIntent);
+
+    // Start network provider along with gps
+    if (shouldStartNetworkProvider(request.getPriority())) {
+      try {
+        locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, request.getInterval(),
+                request.getDisplacemnt(), pendingIntent);
+      } catch (IllegalArgumentException iae) {
+        iae.printStackTrace();
+      }
+    }
+
+    // TODO: should we broadcast last location?
```

Comment:
IMHO it would be good and consistent to have the same behavior with the previous one, which broadcasts it, too

Reply:
agreed. this is intentional, as i'm not sure how `lastLocation` will impact user experience: my goal would be to run this engine both in background and foreground mode and analyze the diff in the experience.